### PR TITLE
✨ Add ability to provision Windows VMs

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
Upstream changes from:
- https://github.com/openshift/cluster-api-provider-azure/pull/140
- https://github.com/openshift/cluster-api-provider-azure/pull/164

 

 
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR provides alternative way to bring Windows VMs in CAPZ.

The functionality allows to bring Windows machines and configure SSH access via `customData` script.

This PR allows to pass `UnattendConfig` to Windows machines on Azure for autostarting custom bootstrap data 
Currently on Azure, the `CustomData` (also called bootstrap data) is written to `C:\AzureData\CustomData.bin`, but nothing will run it.

With this change, Windows machines on Azure will automatically execute the `CustomData` script if it exists.

This is the same behavior as Windows machines on AWS show.

This way of bootstrapping a machine with a provided script in the CustomData (e.g. to setup SSH access, and the kubelet) is leveraged by the [WMCO](https://github.com/openshift/windows-machine-config-operator) and does not require the kubeadm bootstrapping mechanism nor creating custom Windows OS images.

The PR is not requiring amendments to the [proposal](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1035#issuecomment-728258731) due to no plans in using kubeadm bootstrap provider.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/153 for WMCO.

**Special notes for your reviewer**:

cc @aravindhp @LorbusChris please review

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by jsturtevant** [comment]
Thanks!  I've been working on [this](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/153#issuecomment-698643137)  along with a [PR in image builder](https://github.com/kubernetes-sigs/image-builder/pull/382).  My wip (which I haven't opened yet as I was working out some e2e scenarios) is a bit further along. I have e2e tests and such. I will get it opened up so you can take a look and comment?

The one thing I did was a bit different on setting up the SSH.  What is your init script that your passing along?

**Comment by aravindhp** [comment]
> What is your init script that your passing along?

@jsturtevant [this](https://github.com/openshift/windows-machine-config-operator/blob/master/pkg/controller/secrets/secrets.go#L62) is what ends up in the script

**Comment by jsturtevant** [comment]
Interesting so we would need to use the Operator to deploy machines?

As mentioned, I have shared the in progress work, which is pretty similar at the base level of adding OS profile info: #1036. It deploys images as is but still working through the polish.

**Comment by aravindhp** [comment]
>Interesting so we would need to use the Operator to deploy machines?

@jsturtevant on OpenShift, the [Machine API Operator](https://github.com/openshift/machine-api-operator) and its operands handles MachineSets and Machines. The [Windows Machine Config Operator (WMCO)](https://github.com/openshift/windows-machine-config-operator) watches for Windows Machines and configures them so that they join the cluster as a worker node. Ref: [WMCO design](https://github.com/openshift/enhancements/blob/master/enhancements/windows-containers/windows-machine-config-operator.md#design-details).

>As mentioned, I have shared the in progress work, which is pretty similar at the base level of adding OS profile info: #1036. It deploys images as is but still working through the polish.

I will defer to @Danil-Grigorev on this.

**Comment by jsturtevant** [comment]
I think we should be able to make both approaches work.  I don't think there has been any discussion on taking a dependency on the operators but it makes sense that we could support operators and non-operator based scenarios.

File: cloud/services/virtualmachines/virtualmachines.go [hunk 1]
```diff
@@ -39,6 +41,34 @@ import (
...
+			<Enabled>true</Enabled>
+			<LogonCount>1</LogonCount>
+		</AutoLogon>`
+
+	// winFirstLogonCommandsString is the string used to create the FirstLogonCommands
+	// AdditionalUnattendContent configuration for Windows machines.
+	winFirstLogonCommandsString = `<FirstLogonCommands>
+			<SynchronousCommand>
+				<Description>Copy user data secret contents to init script</Description>
+				<CommandLine>cmd /c "copy C:\AzureData\CustomData.bin C:\init.ps1"</CommandLine>
```
Comment (LorbusChris) [inline cloud/services/virtualmachines/virtualmachines.go:61]: per https://github.com/Azure/azure-sdk-for-go/blob/master/services/compute/mgmt/2020-06-01/compute/models.go#L7202-L7203 it seems passing CustomData is not required (if it does, it'll be written to `C:\AzureData\CustomData.bin`).
I reckon these commands here should first check whether the file exists, before copying them (or executing them in the second `SynchronousCommand`. (This check wasn't necessary for us downstream, as our OpenShift machine-api-operator would _always_ pass it iirc)

**Comment by jsturtevant** [comment]
Looking into it a bit further, it looks like this approach uses the new windows specific bootstrap provider [WMCB](https://github.com/openshift/windows-machine-config-bootstrapper) along with the [WMCO](https://github.com/openshift/windows-machine-config-operator).  Is that right?

**Comment by LorbusChris** [comment]
The WMCO is in charge of operating the machines (owning the config that is written as CustomData, as well as other things like starting and stopping services), but the actual provisioning is done with the machine-api-operator, which levarages this custer-api-provider here. The ability to provision a Windows machine should exist with this change here, without the WMCO.

The WMCB on the other hand is in charge of setting up the kubelet so it joins the cluster as a node. There is no equivalent to that in this PR, although one could probably script that purely with the contents of the CustomData script.

**Comment by CecileRobertMichon** [comment]
@Danil-Grigorev @aravindhp @LorbusChris thanks for the PR and welcome to CAPZ! As James mentioned there is already ongoing work to support windows worker nodes in Cluster API (all infra providers that support it, not just Azure) based on [this proposal](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200804-windows-support.md). #1036 implements the proposal for CAPZ (together with https://github.com/kubernetes-sigs/image-builder/pull/382). 

If you are interested in extending the Cluster API Windows support to other bootstrap providers (especially external ones like https://github.com/openshift/windows-machine-config-bootstrapper), I suggest opening an amendment to the proposal above for v1alpha4 (proposals currently being accepted). You can also join the next community [office hours](https://github.com/kubernetes/community/blob/master/sig-cluster-lifecycle/README.md#cluster-api) so we can discuss further (this week is kubecon and next week is US Thanksgiving so next ones will most likely be in two weeks). 

cc @vincepri @devigned @nader-ziada

**Comment by michmike** [comment]
@jayunit100 can you please take a look as well?

**Comment by devigned** [comment]
Just wanted to give a gentle nudge. Is there interest in amending the above linked proposal with the goal of continuing this feature?

**Comment by aravindhp** [comment]
@Danil-Grigorev a gentle ping on this. Any chance you could rebase and update the description as discussed during the cluster API call?

**Comment by jsturtevant** [comment]
There is a new CAEP that might be applicable and make it easier to support both types of bootstrappers https://github.com/kubernetes-sigs/cluster-api/pull/4221

**Comment by aravindhp** [comment]
> There is a new CAEP that might be applicable and make it easier to support both types of bootstrappers [kubernetes-sigs/cluster-api#4221](https://github.com/kubernetes-sigs/cluster-api/pull/4221)

I am not sure sure, @jsturtevant. This would still require customers to prep golden Windows images with the required bits. The [OpenShift Windows Machine Config Operator](https://github.com/openshift/windows-machine-config-operator) is trying not to have customers create custom images for it to be used as worker nodes at least in public clouds. This lowers the barrier to entry for Windows workloads.

**Comment by jsturtevant** [comment]
haven't read through it fully yet but you should leave some comments to see if your use case can be accommodated.

**Comment by LorbusChris** [comment]
> There is a new CAEP that might be applicable and make it easier to support both types of bootstrappers [kubernetes-sigs/cluster-api#4221](https://github.com/kubernetes-sigs/cluster-api/pull/4221)

@jsturtevant this change doesn't try to add any full-blown alternative bootstrapper - the only goal is for the CustomData script to be run automatically on first boot - this is what Windows Machines on AWS do, too.

This change should be entirely orthogonal to the kubeadm bootstrapping mechanism, which we don't plan on using at all at this time, so I don't think it makes sense to mingle this with https://github.com/kubernetes-sigs/cluster-api/pull/4221.

**Comment by devigned** [comment]
@Danil-Grigorev is this still an active work in progress, or should this PR be closed?

**Comment by LorbusChris** [comment]
I've opened https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1433 which is intended to supersede this PR

**Comment by devigned** [comment]
/close

I'm going to close this PR as #1433 is intended to supersede this one.

**Pull request closed without merging**